### PR TITLE
Writing to compressed files / connections

### LINKED ISSUE #348: Writing compressed files
Is there currently any way to write to a compressed text file?

Ideally, compression would be automatically detected from the `path` (ending in `.gz`, `.bz2`, or `.xz`) argument of `readr::write_tsv()`.

### PR DESCRIPTION
**Pull request by jimhester**
Fixes #348 

This still needs documentation, more tests and examples, but works pretty well.

The connection writing uses [boost::iostreams](http://www.boost.org/doc/libs/1_63_0/libs/iostreams/doc/), so is nicely buffered and should be pretty performant.

### CODE REVIEW

File: tests/testthat/test-write-delim.R [hunk 1]
```diff
@@ -92,3 +92,12 @@ test_that("does not writes a tailing .0 for whole number doubles", {
 
   expect_equal(format_tsv(tibble::data_frame(x = -123456789)), "x\n-123456789\n")
 })
+
+test_that("write_csv can write to compressed files", {
+  mt <- read_csv(readr_example("mtcars.csv.bz2"))
+
+  filename <- file.path(tempdir(), "mtcars.csv.bz2")
+  on.exit(unlink(filename))
+  write_csv(mt, filename)
```
Comment (hadley) [inline tests/testthat/test-write-delim.R]: What turns this path into a `bzfile()` connection?

Comment (jimhester) [inline tests/testthat/test-write-delim.R]: https://github.com/tidyverse/readr/pull/609/files#diff-fd81d97b84af37e70673ba40af498d75R144

Comment (hadley) [inline tests/testthat/test-write-delim.R]: I still don't get it 😦  Shouldn't there be a switch on the file extension that creates a connection given a path?

Comment (hadley) [inline tests/testthat/test-write-delim.R]: Oooooh it's standardise path. Duh

**Comment by hadley** [comment]
It might be worth benchmarking to see the overhead of the connection interface. If it's not high, we could consider pouring everything through connections.

**Comment by jimhester** [comment]
From my tests It seems like the overhead for connections is quite low, so I think we could safely handle all output through connections without trouble.

**Comment by hadley** [comment]
In that case, I'd say go ahead and pour everything through the connection interface. That should become our standard approach.

File: src/write_connection.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
+#include "write_connection.h"
+
+#define class class_name
+#define private private_ptr
+#include <R_ext/Connections.h>
+#undef class
+#undef private
+
+#if R_CONNECTIONS_VERSION != 1
```
Comment (jimhester) [inline src/write_connection.cpp]: It was added in https://github.com/wch/r-source/commit/41b64b12f36851cb87e9132612591d623c217443, and running `git branch --contains 41b64b12f36851cb87e91326125` on the r-source clone returns

```
  R-3-3-branch
  build
  build-trunk
  tags/R-3-0-3
  tags/R-3-1-3
  tags/R-3-2-3
  tags/R-3-2-4
  tags/R-3-2-4-revised
  tags/R-3-2-5
  tags/R-3-3-0
  tags/R-3-3-1
  tags/R-3-3-2
  travis
* trunk
```

So looks like the earliest release was 3.0.3

File: tests/testthat/test-write-delim.R (hunk 1 above)
Comment (jimhester) [inline tests/testthat/test-write-delim.R]: Yeah that is a good point, I will think about a test to verify it is a bz2 file, probably the best way would be to check the magic number I think.

File: R/write.R [hunk 3]
```diff
@@ -95,7 +93,7 @@ format_delim <- function(x, delim, na = "NA", append = FALSE, col_names = !appen
   stopifnot(is.data.frame(x))
 
   x <- lapply(x, output_column)
-  stream_delim(x, "", delim, col_names = col_names, append = append, na = na)
+  stream_delim(x, NULL, delim, col_names = col_names, append = append, na = na)
```
Comment (jimhester) [inline R/write.R]: Unfortunately not, `stdout()` is a text-mode connection and the connections API only works with binary mode connections, you get `'write' not enabled for this connection` errors if you try to use `stdout()` here.

Comment (hadley) [inline R/write.R]: Oh yeah :(

File: src/write_connection.cpp (hunk 2 above)
Comment (hadley) [inline src/write_connection.cpp]: Ok, awesome

File: R/source.R [hunk 4]
```diff
@@ -102,29 +102,41 @@ read_connection <- function(con) {
...
+standardise_path <- function(path, input = TRUE) {
   if (!is.character(path))
     return(path)
 
   if (grepl("\n", path))
     return(path)
 
   if (is_url(path)) {
+    if (requireNamespace("curl", quietly = TRUE)) {
+      con <- curl::curl(path)
```
Comment (hadley) [inline R/source.R]: Bump

Comment (jimhester) [inline R/source.R]: I put the message below for the fallback, you are saying you want the message if we are using curl instead?

Comment (hadley) [inline R/source.R]: Sorry, I missed that because of the stupid issues threading. All looks good :)

**Pull request merged**